pkg/cloud/services/ec2/ami.go

```diff
@@ -94,8 +97,51 @@ func GenerateAmiName(amiNameFormat, baseOS, kubernetesVersion string) (string, e
 	return templateBytes.String(), nil
 }
 
+func containsArch(archList []*string, arch string) bool {
+	for _, item := range archList {
+		if *item == arch {
+			return true
+		}
+	}
+	return false
+}
+
+// Determine architecture based on instance type.
+func (s *Service) pickArchitectureForInstanceType(instanceType string) (string, error) {
+	descInstanceTypeInput := &ec2.DescribeInstanceTypesInput{
+		InstanceTypes: []*string{&instanceType},
+	}
+	describeInstanceTypeResult, err := s.EC2Client.DescribeInstanceTypes(descInstanceTypeInput)
+	if err != nil {
+		return "", err
+	}
+
+	supportedArchs := describeInstanceTypeResult.InstanceTypes[0].ProcessorInfo.SupportedArchitectures
+
+	logger := s.scope.GetLogger().WithValues("instance type", instanceType, "supported architectures", supportedArchs)
+	logger.Info("Obtained a list of supported architectures for instance type")
+
+	architecture := ""
+	switch len(supportedArchs) {
+	case 1:
+		architecture = *(supportedArchs[0])
```

Comment:
What would be `supportedArchs[0]` in this case? why is it distinguished based on the length of `supportedArchs`?

Reply:
This was something I took from the old PR, I've just pushed a commit to change how this is handled. Instead of doing this elaborate check I've made it more explicit and easier to read.
This does mean if in the future more architecture types are added (maybe `riscv64`?) we will have to manually update this switch statement, which I don't think is a big deal.

A little context on why it's checking the length - this exists is because some instance types have multiple supported architectures. The only notable case I could find is where `i386` and `x86_64` are both supported, it also only seems to be old `t2.` instance types that have this issue.

Reply:
Ok great that makes sense to me.